### PR TITLE
Added tests using sendmany and createmultisig, and some refactoring

### PR DESCRIPTION
**Pull request by stckwok**

### CODE REVIEW

File: test/basic_interop.py [hunk 1]
```diff
@@ -120,11 +121,15 @@ def Test():
...
-    # "--tmpdir=/ramdisk/test",
-    #t.main(["--nocleanup", "--noshutdown"], bitcoinConf, None)
-    eachConf = [bitcoinConf]*4
-    # eachConf[3]["maxlimitertxfee"] = None  # classic does not have this option
-    t.main(["--tmpdir=/ramdisk/test"], bitcoinConf, None)
+    # folder to store bitcoin runtime data and logs
+    tmpdir = "--tmpdir=/tmp/cashInterop"
+    
+    for arg in sys.argv[1:]:
+        if (("--tmpdir=" or "--tmpdir =") in arg):
```
Comment (gandrewstone) [inline test/basic_interop.py]: Don't use the biggest enclosing parens.
--tmpdir<space>= should not ever match because sys.argv has already been split by spaces.  Don't worry about the --tmpdir=<space> case, its not a valid option format

File: test/trans_interop.py [hunk 2]
```diff
@@ -30,7 +90,7 @@ def verify_transaction_amount_confirm(self, nodeTxOut, nodeTxIn, amount):
         nodeTxOut : node Id of the TxOut (debit)
         nodeTxIn : node Id of the TxIn (credit)
         amount : amount to spend in BTC
-    """
+    """
```
Comment (gandrewstone) [inline test/trans_interop.py]: do a "git diff > diff.txt" before you submit a PR and review it for accidental unnecessary changes like the above.

File: test/trans_interop.py [hunk 3]
```diff
@@ -136,21 +261,32 @@ def run_test(self):
...
+        verify_sendto_different_node(self, 0.5)
         
-        # Simple send amount = 10 from Node 1 to 2 
-        verify_transaction_amount_confirm(self, 1, 2, 10)
+        print("Verify transaction amounts on the same nodes")
+        verify_sendto_same_node(self, 0.5)
         
-        # Simple send amount = 10 from Node 1 to 3 
-        verify_transaction_amount_confirm(self, 1, 3, 10)
+        time.sleep(10)
```
Comment (gandrewstone) [inline test/trans_interop.py]: why is a sleep needed here?  sleeping is dangerous because different systems need different sleep times.  Its better to explicitly check for whatever condition you are waiting for.

File: test/trans_interop.py [hunk 4]
```diff
@@ -22,6 +22,66 @@
...
+    See Bitcoin API JSON-RPC:
+    > https://chainquery.com/bitcoin-api/sendmany
+    > https://chainquery.com/bitcoin-api/creatmultisig
+    
+    Input:
+        nodeTxOut : node Id of the TxOut (debit)
+        nodeTxIn : node Id of the TxIn (credit)
+        num_signature : number of signatures required
+    """ 
+    print(" ")
```
Comment (gandrewstone) [inline test/trans_interop.py]: unnecessary print?

File: test/trans_interop.py [hunk 5]
```diff
@@ -39,40 +99,105 @@ def verify_transaction_amount_confirm(self, nodeTxOut, nodeTxIn, amount):
...
-    assert_array_result(self.nodes[nodeTxOut].listtransactions(),
-                        {"txid":txid},
-                        {"category":"send","account":"","amount":Decimal(str(-amount)),"confirmations":0})                 
-    assert_array_result(self.nodes[nodeTxIn].listtransactions(),
-                        {"txid":txid},
-                        {"category":"receive","account":"","amount":Decimal(str(amount)),"confirmations":0})
+
+    if (nodeTxOut != nodeTxIn):
+        #print(self.nodes[nodeTxOut].listtransactions())
+        print("### Balance (Before) : %s" % [ x.getinfo()["balance"] for x in self.nodes])
```
Comment (gandrewstone) [inline test/trans_interop.py]: let's use the logging.info("string") rather than print.  This will let us turn it off during automated testing.

File: test/trans_interop.py (hunk 2 above)
Comment (stckwok) [reply test/trans_interop.py]: thanks

File: test/trans_interop.py (hunk 3 above)
Comment (stckwok) [reply test/trans_interop.py]: Yes, this sleep is not needed. I forgot to remove it.
Retested ok with calling verify_transaction_sendmany(self, 0, 1, 1) just once within this test case.

But when I call this function 3 times with different arguments as shown, 
verify_transaction_sendmany(self, 0, 1, 1)
verify_transaction_sendmany(self, 1, 2, 1)
verify_transaction_sendmany(self, 2, 3, 1)

I keep getting JSONRPC error: importaddress "address" ( "label" rescan ) as shown below.
The reason I added the sleep(10) is because the log shows "this call can take minutes to complete if rescan is true."  It seems to help, but still fail the same way sometimes.

I think I should be able to call this function "verify_transaction_sendmany" many times and with different input combo. 

Have you seen this problem before? Is there anything I have missed or I am not testing it correctly.

Thanks


------------------ error log ---------------------------------------------------------------
JSONRPC error: importaddress "address" ( "label" rescan )

Adds an address or script (in hex) that can be watched as if it were in your wallet but cannot be used to spend.

Arguments:
1. "address"          (string, required) The address
2. "label"            (string, optional, default="") An optional label
3. rescan               (boolean, optional, default=true) Rescan the wallet for transactions

Note: This call can take minutes to complete if rescan is true.

Examples:

Import an address with rescan
> bitcoin-cli importaddress "myaddress"

Import using a label without rescan
> bitcoin-cli importaddress "myaddress" "testing" false

As a JSON-RPC call
> curl --user myusername --data-binary '{"jsonrpc": "1.0", "id":"curltest", "method": "importaddress", "params": ["myaddress", "testing", false] }' -H 'content-type: text/plain;' http://127.0.0.1:8332/

  File "/home/sk1/Dev/cashInterop/test/test_framework/test_framework.py", line 192, in main
    self.run_test()
  File "./trans_interop.py", line 287, in run_test
    verify_transaction_sendmany(self, 2, 3, 1)
  File "./trans_interop.py", line 75, in verify_transaction_sendmany
    self.nodes[nodeTxOut].importaddress(multisig["redeemScript"], "watchonly", False, True)
  File "/home/sk1/Dev/cashInterop/test/test_framework/coverage.py", line 49, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/sk1/Dev/cashInterop/test/test_framework/authproxy.py", line 165, in __call__
    raise JSONRPCException(response['error'])
> /home/sk1/Dev/cashInterop/test/test_framework/authproxy.py(165)__call__()
-> raise JSONRPCException(response['error'])
(Pdb)

File: test/trans_interop.py (hunk 4 above)
Comment (stckwok) [reply test/trans_interop.py]: Removed. Thanks

File: test/trans_interop.py (hunk 5 above)
Comment (stckwok) [reply test/trans_interop.py]: yes so much nicer with logging.info
I have replaced all print with logging.info

File: test/basic_interop.py (hunk 1 above)
Comment (stckwok) [reply test/basic_interop.py]: Removed the biggest enclosing parens

I verified the user defined folder can be created properly at /tmp folder
./example_test.py --tmpdir=/tmp/mytestfolder
2018-01-12 15:52:30,170.INFO: # User input : --tmpdir=/tmp/mytestfolder

sk1@sk1-VirtualBox:~$ ls -la /tmp
.....
drwxrwxr-x  2 sk1  sk1   4096 Jan 12 15:52 mytestfolder

File: test/example_test.py [hunk 6]
```diff
@@ -83,11 +83,15 @@ def Test():
...
-    # "--tmpdir=/ramdisk/test",
-    #t.main(["--nocleanup", "--noshutdown"], bitcoinConf, None)
-    eachConf = [bitcoinConf]*4
-    # eachConf[3]["maxlimitertxfee"] = None  # classic does not have this option
-    t.main(["--tmpdir=/ramdisk/test"], bitcoinConf, None)
+    # folder to store bitcoin runtime data and logs
+    tmpdir = "--tmpdir=/tmp/cashInterop"
+    
+    for arg in sys.argv[1:]:
+        if ("--tmpdir=" or "--tmpdir =") in arg:
```
Comment (gandrewstone) [inline test/example_test.py]: Does "--tmpdir =" ever match?  I think it won't because if you run
"./prog.py --tmpdir =foo"
then argv will be:
["prog.py", "--tmpdir", "=foo"]
(it splits by spaces) So it should never match

Comment (stckwok) [reply test/example_test.py]: It's removed. Thanks.

File: test/trans_interop.py (hunk 3 above)
Comment (stckwok) [reply test/trans_interop.py]: clientDirs = ["xt", "xt", "xt", "xt"] 
clientDirs = ["xt", "abc", "bucash", "bucash"]
clientDirs = ["bucash", "abc", "xt", "bucash"] 

The JSONRPC error seems to be caused by XT client 
Whenever XT is used on the list above, I got the same error
